### PR TITLE
Include the requested current state alongside the latest events in the query API.

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/roomserver/query/query.go [hunk 1]
```diff
@@ -12,13 +13,17 @@ import (
...
 	// Returns 0 if the room doesn't exists.
 	// Returns an error if there was a problem talking to the database.
 	RoomNID(roomID string) (types.RoomNID, error)
-	// Lookup event references for the latest events in the room.
+	// Lookup event references for the latest events in the room and the current state snapshot.
 	// Returns an error if there was a problem talking to the database.
-	LatestEventIDs(roomNID types.RoomNID) ([]gomatrixserverlib.EventReference, error)
+	LatestEventIDs(roomNID types.RoomNID) ([]gomatrixserverlib.EventReference, types.StateSnapshotNID, error)
+	// Lookup the Events for a list of numeric event IDs.
+	// Returns a sorted list of events.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/query/query.go]: Sorted by what? NID? Order of the input array?

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 2]
```diff
@@ -182,6 +195,79 @@ func DifferenceBetweeenStateSnapshots(db RoomStateDatabase, oldStateNID, newStat
 	}
 }
 
+// StringTuplesToNumericTuples converts the string state key tuples into numeric IDs
+// If there isn't a numeric ID for either the event type or the event state key then the tuple is discarded.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Why discard rather than return an error?

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 3]
```diff
@@ -182,6 +195,79 @@ func DifferenceBetweeenStateSnapshots(db RoomStateDatabase, oldStateNID, newStat
...
+// LoadStateAtSnapshotForTuples loads the state for a list of event type and state key pairs at a snapshot.
+// This is used when we only want to load a subset of the room state at a snapshot.
+// This is typically the state before an event or the current state of a room.
+// Returns a sorted list of state entries or an error if there was a problem talking to the database.
+func LoadStateAtSnapshotForTuples(db RoomStateDatabase, stateNID types.StateSnapshotNID, stateKeyTuples []types.StateKeyTuple) ([]types.StateEntry, error) {
+	stateBlockNIDLists, err := db.StateBlockNIDs([]types.StateSnapshotNID{stateNID})
+	if err != nil {
+		return nil, err
+	}
+	stateBlockNIDList := stateBlockNIDLists[0]
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Please explain.

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 4]
```diff
@@ -182,6 +195,79 @@ func DifferenceBetweeenStateSnapshots(db RoomStateDatabase, oldStateNID, newStat
...
+
+	stateEntryLists, err := db.StateEntriesForTuples(stateBlockNIDList.StateBlockNIDs, stateKeyTuples)
+	if err != nil {
+		return nil, err
+	}
+	stateEntriesMap := stateEntryListMap(stateEntryLists)
+
+	// Combined all the state entries for this snapshot.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: > Combined

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 5]
```diff
@@ -182,6 +195,79 @@ func DifferenceBetweeenStateSnapshots(db RoomStateDatabase, oldStateNID, newStat
...
+	// The order of state block NIDs in the list tells us the order to combine them in.
+	var fullState []types.StateEntry
+	for _, stateBlockNID := range stateBlockNIDList.StateBlockNIDs {
+		entries, ok := stateEntriesMap.lookup(stateBlockNID)
+		if !ok {
+			// If the block is missing from the map it means that none of its entries matched a requested tuple.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: And this is non-fatal how?

File: src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go [hunk 6]
```diff
@@ -80,6 +88,7 @@ func (s *eventTypeStatements) prepare(db *sql.DB) (err error) {
 	return statementList{
 		{&s.insertEventTypeNIDStmt, insertEventTypeNIDSQL},
 		{&s.selectEventTypeNIDStmt, selectEventTypeNIDSQL},
+		{&s.bulkSelectEventTypeNIDStmt, bulkSelectEventTypeNIDSQL},
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go]: :star: Much shorter now :D

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go [hunk 7]
```diff
@@ -131,3 +148,99 @@ func (s *stateBlockStatements) bulkSelectStateDataEntries(stateBlockNIDs []types
 	}
 	return results, nil
 }
+
+func (s *stateBlockStatements) bulkSelectFilteredStateBlockEntries(
+	stateBlockNIDs []types.StateBlockNID, stateKeyTuples []types.StateKeyTuple,
+) ([]types.StateEntryList, error) {
+	tuples := stateKeyTupleSorter(stateKeyTuples)
+	sort.Sort(tuples)
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Why?

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go [hunk 8]
```diff
@@ -131,3 +148,99 @@ func (s *stateBlockStatements) bulkSelectStateDataEntries(stateBlockNIDs []types
...
+		if !tuples.contains(entry.StateKeyTuple) {
+			// The select will return the cross product of types and state keys.
+			// So we need to check if type of the entry is in the list.
+			continue
+		}
+
+		if current == nil || types.StateBlockNID(stateBlockNID) != current.StateBlockNID {
+			// The state entry row is for a different state data block to the current one.
+			// So we start appending to the next entry in the list.
+			current = &results[i]
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: I understand what you're doing, but it's not very clear. We're going to return a slice of `results` and we pre-allocate the size of the array because we know that is the upper bound, but we may return fewer than this. This grabs a pointer to one of the zero-initialised `StateEntryList` structs and populates it, then we slice at the end.

It's all rather complicated and gnarly when this could easily be done as a simple:
```go
var results []types.StateEntryList
results = append(results, current)
return results
```

In fact, the *only* downside to this I see is that you can get unlucky on the backing array resizing, and so waste a little more memory, but if this is the Query API then we're just going to send the bytes down the wire then forget this anyway, right? So it'll be GCed pretty rapidly.

tl;dr this feels like a premature optimisation that is harming readability.

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go [hunk 9]
```diff
@@ -131,3 +148,99 @@ func (s *stateBlockStatements) bulkSelectStateDataEntries(stateBlockNIDs []types
...
+
+func stateBlockNIDsAsArray(stateBlockNIDs []types.StateBlockNID) pq.Int64Array {
+	nids := make([]int64, len(stateBlockNIDs))
+	for i := range stateBlockNIDs {
+		nids[i] = int64(stateBlockNIDs[i])
+	}
+	return pq.Int64Array(nids)
+}
+
+type stateKeyTupleSorter []types.StateKeyTuple
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Tests please.

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: I've combined StringTuplesToNumericTuples and LoadStateAtSnapshotForNumericTuples into a single function. This should make it clearer why we discard unknown event-types and state keys.

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 3 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Done 1fe07fa

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 4 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Fixed 066f9da

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 5 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Done 322ee66

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Better, thank you!

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go (hunk 8 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: I've rewritten it to use `append` c528ce4

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go [hunk 10]
```diff
@@ -131,3 +148,106 @@ func (s *stateBlockStatements) bulkSelectStateDataEntries(stateBlockNIDs []types
...
+			current.StateEntries = nil
+		}
+		current.StateEntries = append(current.StateEntries, entry)
+	}
+	// Add the last entry to the list if it is not empty.
+	if current.StateEntries != nil {
+		results = append(results, current)
+	}
+	// Because we have filtered the list it's possible that some of the blocks were completely removed
+	// from the result.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: This comment is now redundant.

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go [hunk 11]
```diff
@@ -131,3 +148,106 @@ func (s *stateBlockStatements) bulkSelectStateDataEntries(stateBlockNIDs []types
...
+
+		if types.StateBlockNID(stateBlockNID) != current.StateBlockNID {
+			// The state entry row is for a different state data block to the current one.
+			// So we append the current entry to the results and start adding to a new one.
+			// The first time through the loop current will be empty.
+			if current.StateEntries != nil {
+				results = append(results, current)
+			}
+			current.StateBlockNID = types.StateBlockNID(stateBlockNID)
+			current.StateEntries = nil
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Why are we resetting the old `current` rather than making a new `current`?

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go (hunk 9 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Done 0986dd2

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go (hunk 10 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Done dd45dda

File: src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go (hunk 11 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/state_block_table.go]: Fixed 83e20be

File: src/github.com/matrix-org/dendrite/roomserver/query/query.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/query/query.go]: Commented e6bd741

**Comment by Kegsay** [comment]
LGTM after conflict resolution.
